pkg/runtime/serializer/recognizer/recognizer.go

```diff
@@ -27,50 +28,110 @@ import (
+// 2. All other decoders, and any decoder that returned true for unknown.
+//
+// The order passed to the constructor is preserved within those priorities.
 func NewDecoder(decoders ...runtime.Decoder) runtime.Decoder {
-	recognizing, blind := []RecognizingDecoder{}, []runtime.Decoder{}
-	for _, d := range decoders {
-		if r, ok := d.(RecognizingDecoder); ok {
-			recognizing = append(recognizing, r)
-		} else {
-			blind = append(blind, d)
-		}
-	}
 	return &decoder{
-		recognizing: recognizing,
-		blind:       blind,
+		decoders: decoders,
 	}
 }
 
 type decoder struct {
-	recognizing []RecognizingDecoder
-	blind       []runtime.Decoder
+	decoders []runtime.Decoder
 }
 
-func (d *decoder) Decode(data []byte, gvk *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error) {
-	var lastErr error
-	for _, r := range d.recognizing {
-		buf := bytes.NewBuffer(data)
-		ok, err := r.RecognizesData(buf)
-		if err != nil {
-			lastErr = err
-			continue
+var _ RecognizingDecoder = &decoder{}
+
+func (d *decoder) RecognizesData(peek io.Reader) (bool, bool, error) {
+	var (
+		lastErr    error
+		anyUnknown bool
+	)
+	data, _ := bufio.NewReaderSize(peek, 1024).Peek(1024)
+	for _, r := range d.decoders {
+		switch t := r.(type) {
+		case RecognizingDecoder:
+			ok, unknown, err := t.RecognizesData(bytes.NewBuffer(data))
+			if err != nil {
+				lastErr = err
+				continue
+			}
+			anyUnknown = anyUnknown || unknown
+			if !ok {
+				continue
+			}
+			return true, false, nil
 		}
-		if !ok {
-			continue
+	}
+	return false, anyUnknown, lastErr
+}
+
+func (d *decoder) Decode(data []byte, gvk *unversioned.GroupVersionKind, into runtime.Object) (runtime.Object, *unversioned.GroupVersionKind, error) {
+	var (
+		lastErr error
+		skipped []runtime.Decoder
+	)
+
+	// try recognizers, record any decoders we need to give a chance later
+	for _, r := range d.decoders {
+		switch t := r.(type) {
+		case RecognizingDecoder:
+			buf := bytes.NewBuffer(data)
+			ok, unknown, err := t.RecognizesData(buf)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+			if unknown {
+				skipped = append(skipped, t)
+				continue
+			}
+			if !ok {
+				continue
+			}
+			return r.Decode(data, gvk, into)
 		}
-		return r.Decode(data, gvk, into)
 	}
-	for _, d := range d.blind {
-		out, actual, err := d.Decode(data, gvk, into)
-		if err != nil {
-			lastErr = err
-			continue
+
+	for _, r := range d.decoders {
+		switch r.(type) {
+		case RecognizingDecoder:
+			// only give skipped decoders a chance
+			for _, s := range skipped {
+				if s != r {
+					continue
+				}
+				out, actual, err := r.Decode(data, gvk, into)
+				if err != nil {
+					lastErr = err
+					continue
+				}
+				return out, actual, nil
+			}
+		default:
+			// if this is a non-RecognizingDecoder
+			out, actual, err := r.Decode(data, gvk, into)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+			return out, actual, nil
```

Comment:
Both branches in this switch seem to be the same.
Wouldn't it be easier to:
1. add 
  default: skipped = append(skipped, t)
  in the previous for loop (the one trying only recognizing decoders)
2. here simply iterate over skilled:
  for _, s := range skipped {
    out, actual, err := s.Decode(data, gvk, into)
   ...

?